### PR TITLE
Do not expand fragments matched zero times

### LINKED ISSUE #961: Expanding fragment matched zero times should succeed
```rust
macro_rules! add_exprs {
    ($($e:expr)?) => ( 0 $(+ $e)?)
}
```
should expand properly to `0` if no argument is given to `add_exprs!`, and to `0 + $e` if one is passed

### PR DESCRIPTION
**Pull request by CohenArthur**
Closes #961

### CODE REVIEW

File: gcc/rust/expand/rust-macro-expand.cc [hunk 1]
```diff
@@ -3735,10 +3763,10 @@ MacroExpander::transcribe_rule (
     = substitute_tokens (invoc_stream, macro_rule_tokens, matched_fragments);
 
   // // handy for debugging
-  // for (auto &tok : substituted_tokens)
-  //   {
-  //     rust_debug ("tok: [%s]", tok->as_string ().c_str ());
-  //   }
+  for (auto &tok : substituted_tokens)
```
Comment (philberty) [inline gcc/rust/expand/rust-macro-expand.cc:3766]: we should probably make a helper at some point to turn this into a simple std::string at some point in another PR

File: gcc/rust/expand/rust-macro-expand.cc [hunk 2]
```diff
@@ -3863,60 +4046,82 @@ MacroExpander::substitute_tokens (
...
+	  for (auto &token : expanded)
+	    replaced_tokens.emplace_back (token->clone_token ());
 	}
       else
 	{
-	  replaced_tokens.push_back (tok->clone_token ());
+	  replaced_tokens.emplace_back (tok->clone_token ());
 	}
+
+      // std::vector<std::unique_ptr<AST::Token>> parsed_toks;
```
Comment (philberty) [inline gcc/rust/expand/rust-macro-expand.cc:4078]: I would suggest removing this when your done

**Review by philberty — APPROVED** [review]

File: gcc/rust/expand/rust-macro-expand.cc [hunk 3]
```diff
@@ -3855,6 +3883,161 @@ MacroExpander::transcribe_rule (
...
+	rust_debug ("expanding repetition");
+	std::vector<std::unique_ptr<AST::Token>> repetition_pattern;
+	size_t pattern_start = token_idx + 1;
+	size_t pattern_end = pattern_start;
+	for (; pattern_end < macro.size ()
+	       && macro.at (pattern_end)->get_id () != RIGHT_PAREN;
+	     pattern_end++)
+	  ;
+
+	// FIXME: This skips whitespaces... Is that okay??
```
[the remark this answers is not in the document]
Comment (CohenArthur) [reply gcc/rust/expand/rust-macro-expand.cc:4009]: I think it should be fine for now, I'll open up issues. The repetition PR is already going to be reaaaaaally big and without complex test cases which would require this. But I'll definitely work on it :)

**Comment by CohenArthur** [comment]
Closing this to add it to my branch on my fork #956

**Pull request closed without merging**